containers/src/Data/Sequence/Internal.hs

```diff
@@ -1376,6 +1377,210 @@ applicativeTree n !mSize m = case n of
     deepA = liftA3 (Deep (n * mSize))
     emptyTree = pure EmptyT
 
+data RCountMid a = RCountMid
+  !Int  --total size
+  !(Node a)  -- End of the first
+  !Int -- Number of units in the middle
+  !(Node a)  -- Beginning of the last
+
+{-
+We could generalize beforeSeq quite easily to
+
+  beforeSeq :: (a -> c) -> Seq a -> Seq b -> Seq c
+
+This would let us add a rewrite rule
+
+  fmap f xs <* ys  ==>  beforeSeq f xs ys
+
+We don't currently bother because I don't yet know of a practical use for (<*)
+for sequences; a rewrite rule to optimize it seems like extreme overkill.
+-}
+
+beforeSeq :: Seq a -> Seq b -> Seq a
+beforeSeq xs ys = beforeSeq' xs (length ys)
+
+beforeSeq' :: Seq a -> Int -> Seq a
```

Comment:
It would probably be a good idea to comment on the basic ideas somewhere around here... from a 2-3 tree point of view, the result is obtained by fixing a 2-3 tree shape of size `lenys >= 2` (which can be constructed in logarithmic time and space) that is then planeted at every leaf of `xs`.

The shape is defined recursively in a bottom-up fashion; at each level there is an initial node, a middle node that has to be replicated a number of times, and a final node, `(l, n x m, r)`. For n = 0 or n = 1, a single node is enough to accomplish this; for n = 2, two nodes are required. In all other cases, we combine l with one or two m's into a single node and symmetrically, one or two m's with r for the last node, such that the remaining number of m's (n-4 to n-2) is divisible by 3. So for the middle part, we can combine 3 m's and then recurse.

The code is complicated by the fact that the first and last elements of the initial list are put on the fringes of the finger tree while the shape is being fixed. This involves quite a bit of bookkeeping (partly captured by `RCountMid` above).

Reply:
Yes, an explanation of the basic ideas is indeed in order. I don't yet understand the second paragraph in your comment above: specifically what you mean about "a middle node that has to be replicated a number of times". I was thinking about the general process as using a (virtual) rigid tree for the second argument (the one where we only care about the length) in which every node that's not on the spine is a 3-node. To create the corresponding 2-3 tree, we need to create three things: the squashed-up left spine, the squashed-up right spine, and (in case we have a `Single` at the bottom) a full 3-tree for the center. Fortunately, that full 3-tree is produced "for free" out of the process of building the other parts, since we need such trees at each level to fill in the internal nodes. But I know my thinking about this is rather fuzzy in general; I mostly worked things out by trial and error starting with the geometrical intuition. So it's very likely possible to explain it better some other way. But yours remains mysterious to me.

Reply:
Well, your perspective is different from mine. In my mind, I build the full 2-3 tree first. The finger tree is then obtained by peeling off digits from the left and the right, leaving leftover digits of size 1 or 2. (This can be done on the fly as the 2-3 tree is being built.) Btw, my _n_ corresponds to `deep_count` in your code.

Anyway, let's see an example. To build a full 2-3 tree with 14 nodes, we can combine them as follows:

```
abbbbbbbbbbbbc -- 14 nodes: combine triples in the middle, and pairs or triples at the ends
(ab)(bbb)(bbb)(bbb)(bbc) -- 4 nodes: ditto
((ab)(bbb))((bbb)(bbb)(bbc)) -- 2 nodes: base case
(((ab)(bbb))((bbb)(bbb)(bbc)))
```

The digits that can be peeled off are `a,b` and `(bbb)` from the left, and `b,b,c` and `(bbb),(bbb)` from the right.